builtin/reset.c

```diff
@@ -162,7 +163,12 @@ static void update_index_from_diff(struct diff_queue_struct *q,
 					die(_("make_cache_entry failed for path '%s'"),
 						two->path);
 
-				checkout_entry(ceBefore, &state, NULL);
+				if (file_exists(two->path)) {
```

Comment:
It looks like this logic is very specific to GVFS and our knowledge of how it behaves.  As such, it should be wrapped with the appropriate gvfs_config_is_set() test.

The rest should probably be submitted upstream as it improves a boundary case that exists for regular sparse checkout repos.

Reply:
As discussed offline, rather than opening the file for write can we try clearing the file's skip-worktree bit?  VFS for Git should pick that change up and automatically add the path to ModifiedPaths.dat.